compiler/rustc_codegen_ssa/src/glue.rs

```diff
@@ -59,37 +59,41 @@ pub fn size_and_align_of_dst<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>>(
             debug!("DST {} layout: {:?}", t, layout);
 
             let i = layout.fields.count() - 1;
-            let sized_size = layout.fields.offset(i).bytes();
+            let unsized_offset_unadjusted = layout.fields.offset(i).bytes();
             let sized_align = layout.align.abi.bytes();
-            debug!("DST {} statically sized prefix size: {} align: {}", t, sized_size, sized_align);
-            let sized_size = bx.const_usize(sized_size);
+            debug!(
+                "DST {} offset of dyn field: {}, statically sized align: {}",
+                t, unsized_offset_unadjusted, sized_align
+            );
+            let unsized_offset_unadjusted = bx.const_usize(unsized_offset_unadjusted);
             let sized_align = bx.const_usize(sized_align);
 
             // Recurse to get the size of the dynamically sized field (must be
             // the last field).
             let field_ty = layout.field(bx, i).ty;
             let (unsized_size, mut unsized_align) = size_and_align_of_dst(bx, field_ty, info);
 
-            // FIXME (#26403, #27023): We should be adding padding
-            // to `sized_size` (to accommodate the `unsized_align`
-            // required of the unsized field that follows) before
-            // summing it with `sized_size`. (Note that since #26403
-            // is unfixed, we do not yet add the necessary padding
-            // here. But this is where the add would go.)
+            // # First compute the dynamic alignment
 
-            // Return the sum of sizes and max of aligns.
-            let size = bx.add(sized_size, unsized_size);
-
-            // Packed types ignore the alignment of their fields.
+            // Packed type alignment would have to be capped, but their tails always have known alignment.
+            // Therefore, their alignment has already been taken into account when computing `sized_align`
+            // and `unsized_offset_unadjusted`, so no further adjustment is needed.
             if let ty::Adt(def, _) = t.kind() {
                 if def.repr().packed() {
+                    let unsized_tail =
+                        bx.tcx().struct_tail_with_normalize(field_ty, |ty| ty, || {});
+                    assert!(matches!(unsized_tail.kind(), ty::Slice(..) | ty::Str));
```

Comment:
I think this is not right. Won't this ICE for code like [this](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=17cef827f7b733af4d0eb7e9cab0ccf1):

```rust
#[repr(C, packed)]
struct Packed<T: ?Sized>(u8, core::mem::ManuallyDrop<T>);

let p = Packed(0, core::mem::ManuallyDrop::new(1));
let p: &Packed<usize> = &p;
let p: &Packed<dyn Send> = p;
dbg!(core::mem::size_of_val(p), core::mem::align_of_val(p));
```
?

Reply:
Hm, if such code is accepted then it would currently just compute the wrong offset for that code. An ICE would still be an improvement. ;)

I'll play around with examples that involve `ManuallyDrop` to see if there's a bug here, thanks!

Reply:
> - There is no justification for why packed types ignore the alignment of their fields. This is correct only because it is impossible to write a packed struct with `dyn Trait` tail, i.e., unsized packed structs always have statically known alignment. (With slices, the size is dynamic but the alignment is static; only with `dyn Trait` do we have both dynamic size and alignment.)

I don't think that's correct, given the example above? And either way, packed structs always ignore the alignment of their fields, don't they? I don't think `Packed<dyn Send>`'s size (or alignment) depends on the alignment of `.0`.

Reply:
> packed structs always ignore the alignment of their fields, don't they? 

No they don't. Remember we have `#[packed(4)]`. The alignment of the fields in that struct is  `min(natural_field_align, 4)`.

Reply:
Right, the alignment should only be ignored if `packedness = 1`. The bug is trivial to trigger then, yikes: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=80f13045af7f1dd52cf6c79ee2897829

Reply:
Yes indeed.

We also forgot to fixup the offset computation logic, I fixed that recently in https://github.com/rust-lang/rust/pull/118540. I thought size computation wouldn't be affected since I was unable to construct a packed struct with a dyn-typed tail, but you managed to bring in that missing piece.

Reply:
I was also made aware of https://github.com/rust-lang/rust/issues/80925, did https://github.com/rust-lang/rust/pull/118540 fix that one? (I can't reproduce miri failure anymore)

Reply:
Yeah for offsets, https://github.com/rust-lang/rust/pull/118540 removed all special cases for packed types, so they are now handled properly.

Reply:
I updated the PR to fix this, and added tests.